File: lib/bindings/langs/android/lib/src/main/kotlin/breez_sdk_liquid_notification/job/LnurlPayVerify.kt
package breez_sdk_liquid_notification.job

import android.content.Context
import breez_sdk_liquid.BindingLiquidSdk
import breez_sdk_liquid.GetPaymentRequest
import breez_sdk_liquid.PaymentDetails
import breez_sdk_liquid.PaymentState
import breez_sdk_liquid_notification.Constants.CACHE_CONTROL_MAX_AGE_THREE_SEC
import breez_sdk_liquid_notification.Constants.CACHE_CONTROL_MAX_AGE_WEEK
import breez_sdk_liquid_notification.Constants.DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.LNURL_PAY_VERIFY_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.NOTIFICATION_CHANNEL_REPLACEABLE
import breez_sdk_liquid_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_liquid_notification.ResourceHelper.Companion.getString
import breez_sdk_liquid_notification.SdkForegroundService
import breez_sdk_liquid_notification.ServiceLogger
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class LnurlVerifyRequest(
    @SerialName("payment_hash") val paymentHash: String,
    @SerialName("reply_url") val replyURL: String,
)

// Serialize the response according to to LUD-21 verify specification:
// https://github.com/lnurl/luds/blob/luds/21.md
@Serializable
data class LnurlPayVerifyResponse(
    val status: String,
    val settled: Boolean,
    val preimage: String?,
    val pr: String,
) {
    constructor(settled: Boolean, preimage: String?, pr: String) : this("OK", settled, preimage, pr)
}

class LnurlPayVerifyJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
) : LnurlPayJob {
    companion object {
        private const val TAG = "LnurlPayVerifyJob"
    }

    override fun start(liquidSDK: BindingLiquidSdk) {
        var request: LnurlVerifyRequest? = null
        try {
            val decoder = Json { ignoreUnknownKeys = true }
            request = decoder.decodeFromString(LnurlVerifyRequest.serializer(), payload)
            // Get the payment by payment hash
            val getPaymentReq = GetPaymentRequest.PaymentHash(request.paymentHash)
            val payment = liquidSDK.getPayment(getPaymentReq) ?: run {
                throw InvalidLnurlPayException("Not found")
            }
            val response = when (val details = payment.details) {
                is PaymentDetails.Lightning -> {
                    // In the case of a Lightning payment, if it's paid via Lightning or MRH,
                    // we can release the preimage
                    val settled = when (payment.status) {
                        // If the payment is pending, we need to check if it's paid via Lightning or MRH
                        PaymentState.PENDING -> details.claimTxId != null
                        PaymentState.COMPLETE -> true
                        else -> false
                    }
                    LnurlPayVerifyResponse(
                        settled,
                        if (settled) details.preimage else null,
                        details.invoice!!,
                    )
                }
                else -> null
            } ?: run {
                throw InvalidLnurlPayException("Not found")
            }

            val maxAge = if (response.settled) CACHE_CONTROL_MAX_AGE_WEEK else CACHE_CONTROL_MAX_AGE_THREE_SEC
            val success = replyServer(Json.encodeToString(response), request.replyURL, maxAge)
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_REPLACEABLE,
                getString(
                    context,
                    if (success) LNURL_PAY_VERIFY_NOTIFICATION_TITLE else LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE,
                    if (success) DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_TITLE else DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE,
                ),
            )
        } catch (e: Exception) {
            logger.log(TAG, "Failed to process lnurl verify: ${e.message}", "WARN")
            if (request != null) {
                fail(e.message, request.replyURL, logger)
            }
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_REPLACEABLE,
                getString(
                    context,
                    LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE,
                    DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE,
                ),
            )
        }

        fgService.onFinished(this)
    }

    override fun onShutdown() {}
}

File: lib/bindings/langs/swift/Sources/BreezSDKLiquid/Task/LnurlPayVerify.swift
import UserNotifications
import Foundation

struct LnurlVerifyRequest: Codable {
    let payment_hash: String
    let reply_url: String
}

struct LnurlVerifyResponse: Decodable, Encodable {
    let status: String
    let settled: Bool
    let preimage: String?
    let pr: String
    
    init(settled: Bool, preimage: String?, pr: String) {
        self.status = "OK"
        self.settled = settled
        self.preimage = preimage
        self.pr = pr
    }
}

class LnurlPayVerifyTask : LnurlPayTask {
    fileprivate let TAG = "LnurlPayVerifyTask"
    
    init(payload: String, logger: ServiceLogger, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        let successNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_VERIFY_NOTIFICATION_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_TITLE)
        let failNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_VERIFY_NOTIFICATION_FAILURE_TITLE)
        super.init(payload: payload, logger: logger, contentHandler: contentHandler, bestAttemptContent: bestAttemptContent, successNotificationTitle: successNotificationTitle, failNotificationTitle: failNotificationTitle)
    }
    
    override func start(liquidSDK: BindingLiquidSdk) throws {
        var request: LnurlVerifyRequest? = nil
        do {
            request = try JSONDecoder().decode(LnurlVerifyRequest.self, from: self.payload.data(using: .utf8)!)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to decode payload: \(e)", level: "ERROR")
            self.displayPushNotification(title: self.failNotificationTitle, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_REPLACEABLE)
            throw e
        }
        
        do {
            // Get the payment by payment hash
            let getPaymentReq = GetPaymentRequest.paymentHash(paymentHash: request!.payment_hash)
            guard let payment = try liquidSDK.getPayment(req: getPaymentReq) else {
                throw InvalidLnurlPayError.notFound
            }
            var response: LnurlVerifyResponse? = nil
            switch payment.details {
                case let .lightning(_, _, _, preimage, invoice, _, _, _, _, _, _, claimTxId, _, _):
                    // In the case of a Lightning payment, if it's paid via Lightning or MRH,
                    // we can release the preimage
                    let settled = switch payment.status {
                        case .pending:
                            // If the payment is pending, we need to check if it's paid via Lightning or MRH
                            claimTxId != nil
                        case .complete:
                            true
                        default:
                            false
                    }
                    response = LnurlVerifyResponse(settled: settled, preimage: settled ? preimage : nil, pr: invoice!)
                default: 
                    break
            }
            if response == nil {
                throw InvalidLnurlPayError.notFound
            }
            let maxAge = response!.settled ? Constants.CACHE_CONTROL_MAX_AGE_WEEK : Constants.CACHE_CONTROL_MAX_AGE_THREE_SEC
            replyServer(encodable: response, replyURL: request!.reply_url, maxAge: maxAge)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to process lnurl verify: \(e)", level: "ERROR")
            fail(withError: e.localizedDescription, replyURL: request!.reply_url)
        }
    }
}
